File: src/frds/measures/modified_merton/mod_merton_create_lookup.py
import numpy as np
from numpy.random import RandomState

from frds.measures.modified_merton.mod_merton_computation import mod_merton_computation


def mod_merton_create_lookup(d, y, T, H, bookD, rho, ltv, xfs, xr, xF, xsig, N, Nsim2):
    # fmt: off
    rng = RandomState(1)
    w = rng.standard_normal((Nsim2, 3 * N))
    
    J = xsig.shape[1]
    K = xr.shape[2]
    Q = xF.shape[3]
    G = xfs.shape[0]
    
    fs = xfs[:, 0, 0, 0] # (1,69,1,105), but in Matlab is (69,1,1,105)
    
    xLt = np.zeros((G, J, K, Q))
    xBt = np.zeros((G, J, K, Q))
    xEt = np.zeros((G, J, K, Q))
    xFt = np.zeros((G, J, K, Q))
    xmdef = np.zeros((G, J, K, Q))
    xsigEt = np.zeros((G, J, K, Q))
    
    last_param = []
    for j in range(J):
        for k in range(K):
            for q in range(Q):
                param = [xr[0, j, k, q], T, xF[0, j, k, q], H, bookD * np.exp(xr[0, j, k, q] * H), rho, ltv, xsig[0, j, k, q], d, y]
                
                # FIXME: outputs dimenstions incorret: only 1*1, should be 69*1
                Lt, Bt, Et, _, _, _, sigEt, mFt, _, mdef, *_ = mod_merton_computation(fs, param, N, Nsim2, w)
                if param!=last_param:
                    print(j,k,q, param, Et)
                    last_param = param
                xLt[:, j, k, q] = Lt.copy()
                xBt[:, j, k, q] = Bt.copy()
                xEt[:, j, k, q] = Et.copy() 
                xFt[:, j, k, q] = mFt.copy()
                xmdef[:, j, k, q] = mdef.copy()
                xsigEt[:, j, k, q] = sigEt.copy()

    return xLt, xBt, xEt, xFt, xmdef, xsigEt # FIXME: somehow all Lt, Bt, Et... remain same for all q

    # fmt: on

File: tests/measures/modified_merton/matlab/ModMertonCreateLookup.m


function [xLt, xBt, xEt, xFt, xmdef, xsigEt] = ModMertonCreateLookup(d,y,T,H,bookD,rho,ltv,xfs,xr,xF,xsig,N,Nsim2) 


rng(1,'twister')
w = normrnd(0,1,[Nsim2, 3*N]);
     
J = size(xsig,2); 
K = size(xr,3); 
Q = size(xF,4); 
G = size(xfs,1); 

fs = xfs(:,1,1,1); 

xLt = zeros(G,J,K,Q); 
xBt = zeros(G,J,K,Q);  
xEt = zeros(G,J,K,Q); 
xFt = zeros(G,J,K,Q); 
xmdef = zeros(G,J,K,Q); 
xsigEt = zeros(G,J,K,Q); 

for j = 1:J
    
    for k = 1:K
        
        for q = 1:Q 
           j
           k
           q
          
           param = [xr(1,j,k,q); T; xF(1,j,k,q); H; bookD*exp(xr(1,j,k,q)*H); rho; ltv; xsig(1,j,k,q); d; y];  %get face value of debt from current book value of debt
        
           [Lt, Bt, Et, ~, ~, ~, sigEt, mFt, ~, mdef] = ModMertonComputation(fs, param, N, Nsim2, w); 
        
           xLt(:,j,k,q) = Lt; 
           xBt(:,j,k,q) = Bt; 
           xEt(:,j,k,q) = Et; 
           xFt(:,j,k,q) = mFt; 
           xmdef(:,j,k,q) = mdef; 
           xsigEt(:,j,k,q) = sigEt; 
           
        end      
    end  
end 


